GoogleChromeLabs/ps-analysis-tool
# Feature: Add the capability of setting "allow list" content policy

## Description

Add the capability of updating the "allow list" Content Policy setting directly from cookies table in PSAT. This PR will add the following functionality

- Ability to right-click on a table row and add a domain to the allow-list.
- The user should then be able to view the allow-listed domains under `chrome://settings/content/siteData`.
- A user should be able to distinguish cookies that have been allow-listed by using colors in the row.
- Ability to remove a domain from the allow-listed settings.
- The allow list setting should only be available during a browser session.

## Relevant Technical Choices

- Use the `[chrome.contentSettings](https://developer.chrome.com/docs/extensions/reference/api/contentSettings) API to add domain to allow list.
- Refactor code in #422 to add allow list functionality from the extension


## Constraints
- There seems to be some issue with the content settings API due to which the cookies which are set as `session_only` are being treated as cookies set to `allow`. The issue has been reported [here](https://bugs.chromium.org/p/chromium/issues/detail?id=1517841 ). Therefore `chrome.storage.session` has been used to store allowed list settings which gets erased after a browser session.

## Testing Instructions

1. Go to a website like https://www.bbc.com/ and analyze the cookies.
2. Right-click on a blocked cookie to open the context menu, and click on "Add domain to allow list".
3. This will add the domain to the allow list, which can be seen as "Blocked Reason (allow listed)" in the bottom description box. However, the yellow color of the blocked cookie will not change, as the cookie is still blocked (unless we refresh the page).
4. If the cookie was blocked due to user preference or 3PCD, it will be allowed and the cookie row will turn green. If the cookie is blocked for any other reason, it will remain blocked with a yellow color row.
5. Once a cookie is added to allowed list, it can also be removed from its context menu as shown in the below screenshot.

Its important to note that if a domain for example .bbc.com is added to allow list, its subdomains like xyz.bbc.com will also be allow listed and if you want to remove it's subdomain from allow list, you would need to remove the parent domain from allow list.

## Screenshot/Screencast

<img width="1283" alt="image" src="https://github.com/GoogleChromeLabs/ps-analysis-tool/assets/6297436/74edc62f-71e2-48f9-850f-ba40809cf3b1">


---

## Checklist






Fixes #334
